arwen/elrondapi/managedConversions.go

```diff
@@ -0,0 +1,275 @@
+	if err != nil {
+		return nil, err
+	}
+	managedType.ConsumeGasForThisIntNumberOfBytes(len(tokenIdentifier))
+	nonce := binary.BigEndian.Uint64(data[4:12])
+	valueHandle := int32(binary.BigEndian.Uint32(data[12:16]))
+	value, err := managedType.GetBigInt(valueHandle)
+	if err != nil {
+		return nil, err
+	}
+	managedType.ConsumeGasForBigIntCopy(value)
+
+	tokenType := core.Fungible
+	if nonce > 0 {
+		tokenType = core.NonFungible
+	}
+
+	return &vmcommon.ESDTTransfer{
+		ESDTTokenName:  tokenIdentifier,
+		ESDTTokenType:  uint32(tokenType),
+		ESDTTokenNonce: nonce,
+		ESDTValue:      value,
+	}, nil
+}
+
+// Converts a managed buffer of serialized data to a slice of ESDTTransfer.
+// The format is:
+// - token identifier handle - 4 bytes
+// - nonce - 8 bytes
+// - value handle - 4 bytes
+// Total: 16 bytes.
+func readESDTTransfers(
+	managedType arwen.ManagedTypesContext,
+	managedVecHandle int32,
+) ([]*vmcommon.ESDTTransfer, error) {
+	managedVecBytes, err := managedType.GetBytes(managedVecHandle)
+	if err != nil {
+		return nil, err
+	}
+	managedType.ConsumeGasForThisIntNumberOfBytes(len(managedVecBytes))
+
+	if len(managedVecBytes)%esdtTransferLen != 0 {
+		return nil, errors.New("invalid managed vector of ESDT transfers")
+	}
+
+	numTransfers := len(managedVecBytes) / esdtTransferLen
+	result := make([]*vmcommon.ESDTTransfer, 0, numTransfers)
+	for i := 0; i < len(managedVecBytes); i += handleLen {
+		esdtTransfer, err := readESDTTransfer(managedType, managedVecBytes[i:i+handleLen])
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, esdtTransfer)
+	}
+
+	return result, nil
+}
+
+// Serializes a vmcommon.ESDTTransfer object.
+func writeESDTTransfer(
+	managedType arwen.ManagedTypesContext,
+	esdtTransfer *vmcommon.ESDTTransfer,
+	destinationBytes []byte,
+) {
+	tokenIdentifierHandle := managedType.NewManagedBufferFromBytes(esdtTransfer.ESDTTokenName)
+	valueHandle := managedType.NewBigInt(esdtTransfer.ESDTValue)
+
+	binary.BigEndian.PutUint32(destinationBytes[0:4], uint32(tokenIdentifierHandle))
+	binary.BigEndian.PutUint64(destinationBytes[4:12], esdtTransfer.ESDTTokenNonce)
+	binary.BigEndian.PutUint32(destinationBytes[12:16], uint32(valueHandle))
+}
+
+// Serializes a list of ESDTTransfer one after the other into a byte slice.
+// The format is (for each ESDTTransfer):
+// - token identifier handle - 4 bytes
+// - nonce - 8 bytes
+// - value handle - 4 bytes
+// Total: 16 bytes.
+func writeESDTTransfersToBytes(
+	managedType arwen.ManagedTypesContext,
+	esdtTransfers []*vmcommon.ESDTTransfer,
+) []byte {
+	destinationBytes := make([]byte, esdtTransferLen*len(esdtTransfers))
+	dataIndex := 0
+	for _, esdtTransfer := range esdtTransfers {
+		writeESDTTransfer(managedType, esdtTransfer, destinationBytes[dataIndex:dataIndex+esdtTransferLen])
+		dataIndex += esdtTransferLen
+	}
+
+	return destinationBytes
+}
+
+type vmInputData struct {
+	destination []byte
+	function    string
```

Comment:
Consider making `function` a byte slice. In `managedei.go` it is always converted from `string` to `[]byte`.

Reply:
it is better to have as string - as you can avoid programming mistakes.